runtime/lua/vim/fswatch/fswatch.lua

```diff
@@ -0,0 +1,140 @@
+--[[
+  Create a file watcher, each watcher is identified by the name of the file that it
+  watches. We can use a lua table to store all watchers indexed by their filenames
+  so that we can close the required watcher during the callback to on_change to
+  debounce the watcher.
+--]]
+
+local uv = vim.loop
+
+local Watcher = {
+  fname = '',
+  ffname = '',
+  handle = nil,
+  paused = false,
+  pending_notifs = false,
+}
+local WatcherList = {}
+
+-- idle handle to check for any pending notifications in a watcher.
+-- Only displays the notifications if neovim is in focus, and the buffer
+-- is the current buffer.
+
+-- Callback for the idle handle
+function check_notifications()
+  for f, watcher in pairs(WatcherList) do
+    if watcher.pending_notifs and watcher.paused == false then
+      if uv.fs_stat(watcher.ffname) ~= nil then
+        vim.api.nvim_command('call fswatch#PromptReload()')
+      else
+        print("ERR: File "..watcher.fname.." removed")
+      end
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+local check_handle = uv.new_check()
+check_handle:start(vim.schedule_wrap(check_notifications))
+
+function Watcher:new(fname)
+  assert(fname ~= '', 'Watcher.new: Error: fname is an empty string')
+  -- get full path name for the file
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {fname = fname, ffname = ffname, handle = nil}
+  setmetatable(w, self)
+  self.__index = self
+  return w
+end
+
+function Watcher:start()
+  assert(self.fname ~= '', 'Watcher.start: Error: no file to watch')
+  assert(self.ffname ~= '', 'Watcher.start: Error: full path for file not available')
+  -- get a new handle
+  self.handle = uv.new_fs_event()
+  self.handle:start(self.ffname, {}, self.on_change)
+end
+
+function Watcher:stop()
+  assert(self.fname ~= '', 'Watcher.stop: Error: no file being watched')
+  assert(self.handle ~= nil, 'Watcher.stop: Error: no handle watching the file')
+  self.handle:stop()
+  -- close the handle altogether, for windows.
+  if self.handle:is_closing() then
+    return
+  end
+  self.handle:close()
+end
+
+function Watcher.on_change(err, fname, event)
+  WatcherList[fname].pending_notifs = true
+
+  WatcherList[fname]:stop()
+  WatcherList[fname]:start()
+end
+
+function Watcher.watch(fname)
+  -- since we can only get file name from callback, use only the file
+  -- name for storing in table. (Without the rest of the path.)
+  local f = vim.api.nvim_call_function('fnamemodify', {fname, ':t'})
+
+  -- if a watcher already exists, close it.
+  if WatcherList[f] ~= nil then
+    WatcherList[f]:stop()
+  end
+
+  -- create a new watcher and it to the watcher list.
+  local w = Watcher:new(fname)
+  WatcherList[f] = w
+  w:start()
+end
+
+function Watcher.stop_watch(fname)
```

Comment:
above function should be called "start_watch" or just "start" and this should be called either "stop_watch" or "stop", depending on what you choose for this function I think.

Reply:
Or perhaps "start_filename_watch" and "stop_filename_watch". It is confusing to have the two sets of "starts/stops"

Reply:
Ah well I too was thinking about it and probably resolved to `start_watch` and `stop_watch` in my latest commit.